src/java/main/org/apache/zookeeper/server/WatchManager.java

```diff
@@ -97,23 +102,44 @@ synchronized void removeWatcher(Watcher watcher) {
     Set<Watcher> triggerWatch(String path, EventType type, Set<Watcher> supress) {
         WatchedEvent e = new WatchedEvent(type,
                 KeeperState.SyncConnected, path);
-        HashSet<Watcher> watchers;
+        Set<Watcher> watchers = new HashSet<>();
+        PathIterator pathIterator = new PathIterator(path);
         synchronized (this) {
-            watchers = watchTable.remove(path);
-            if (watchers == null || watchers.isEmpty()) {
-                if (LOG.isTraceEnabled()) {
-                    ZooTrace.logTraceMessage(LOG,
-                            ZooTrace.EVENT_DELIVERY_TRACE_MASK,
-                            "No watchers for " + path);
+            for (String localPath : pathIterator.asIterable()) {
+                Map<Watcher, Type> thisWatchers = watchTable.get(localPath);
+                if (thisWatchers == null || thisWatchers.isEmpty()) {
+                    continue;
                 }
-                return null;
-            }
-            for (Watcher w : watchers) {
-                HashSet<String> paths = watch2Paths.get(w);
-                if (paths != null) {
-                    paths.remove(path);
+                Iterator<Entry<Watcher, Type>> iterator = thisWatchers.entrySet().iterator();
+                while (iterator.hasNext()) {
+                    Entry<Watcher, Type> entry = iterator.next();
+                    Type entryType = entry.getValue();
+                    Watcher watcher = entry.getKey();
+                    if (entryType == Type.PERSISTENT) {
+                        if ( type != EventType.NodeChildrenChanged ) {
+                            watchers.add(watcher);
```

Comment:
The problem is that it makes 3.5.4 clients incompatible with 3.5.3 servers. That might make sense for 3.6.0 but it's unreasonable for .x.N release.